### PR TITLE
GRPA-3100: Consolidate storage configuration and convert to Butane

### PR DESCRIPTION
**Pull request by bgilbert**
Convert LUKS and mirroring configuration to Butane.  Consolidate the LUKS TPM2, LUKS Tang, and mirroring instructions into a single procedure, since Butane needs to know about the entire storage layout at once.

Update Tang networking advice for current RHCOS features/limitations.

Document the `threshold` field for Shamir's Secret Sharing.

cc @bobfuru @jlebon

Fixes: https://bugzilla.redhat.com/show_bug.cgi?id=1938965

### CODE REVIEW

**Comment by bgilbert** [comment]
cc @arithx

File: modules/installation-special-config-storage.adoc [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+[source,yaml]
+.Butane config example
+----
+variant: openshift
+version: 4.8.0
+metadata:
+  name: worker-storage <1>
+  labels:
+    machineconfiguration.openshift.io/role: worker <1>
+boot_device:
```
Comment (arithx) [inline modules/installation-special-config-storage.adoc:117]: Should probably include the `layout` key as we've mentioned above that this feature is available on `x86_64 nodes booted with BIOS or UEFI and on ppc64le nodes.`

Comment (bgilbert) [reply modules/installation-special-config-storage.adoc:117]: Good catch.  Updated.

**Review by arithx — APPROVED** [review]
Change LGTM

**Comment by bgilbert** [comment]
Updated for comments.

**Review by jlebon — APPROVED** [review]

File: modules/installation-special-config-storage.adoc [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+[source,terminal]
+----
+eyJhbmc3SlRyMXpPenc3ajhEQ01tZVJiTi1oM...
+----
+
+.. If nodes are configured with static IP addressing, ensure you use the coreos-installer `--append-karg` option when installing {op-system} nodes to set the IP address of the installed system.  Append the `ip=` and other arguments needed for your network.
++
+[IMPORTANT]
+====
+Some methods for configuring static IPs do not affect the initramfs after the first boot, and so will not work with Tang encryption.  These include the coreos-installer `--copy-network` option, and adding `ip=` arguments to the kernel command line of the live ISO or PXE image during installation.  Incorrect static IP configuration causes the second boot of the node to fail.
```
Comment (bobfuru) [inline modules/installation-special-config-storage.adoc:95]:
```suggestion
Some methods for configuring static IPs do not affect the initramfs after the first boot and will not work with Tang encryption. These include the `coreos-installer` `--copy-network` option, as well as adding `ip=` arguments to the kernel command line of the live ISO or PXE image during installation. Incorrect static IP configuration causes the second boot of the node to fail.
```

^Tiny style tweaks

File: modules/installation-special-config-storage.adoc [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+====
+
+. Change to the directory that contains the installation program and generate the Kubernetes manifests for the cluster:
++
+[source,terminal]
+----
+$ ./openshift-install create manifests --dir=<installation_directory>
+----
+
+. Create a Butane config configuring disk encryption and/or mirroring. For example, to configure storage for worker nodes, create `$HOME/clusterconfig/worker-storage.bu`:
```
Comment (bobfuru) [inline modules/installation-special-config-storage.adoc:105]:
```suggestion
. Create a Butane config that configures disk encryption, mirroring, or both. For example, to configure storage for worker nodes, create a `$HOME/clusterconfig/worker-storage.bu` file:
```
^For clarity, but please ignore if my changes are incorrect.

File: modules/installation-special-config-storage.adoc [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+    threshold: 1 <6>
+  mirror: <7>
+    devices: <8>
+      - /dev/sda
+      - /dev/sdb
+openshift:
+  fips: true <9>
+----
++
+<1> On master nodes, substitute `master` for `worker` in both of these locations.
```
Comment (bobfuru) [inline modules/installation-special-config-storage.adoc:132]: Similar comments:
"On control plane nodes, substitute `master` (or control plane) for `worker` in both of these locations.

File: modules/installation-special-config-storage.adoc [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+<2> On ppc64le nodes, set this field to `ppc64le`.  On all other nodes, this field can be omitted.
+<3> Include this section if you want to encrypt the root filesystem. See "About disk encryption" for more details.
+<4> Include this field if you want to use a Trusted Platform Module (TPM) to encrypt the root filesystem.
+<5> Include this section if you want to use a Tang server. To obtain the server URL and thumbprint, follow the instructions above.
```
Comment (bobfuru) [inline modules/installation-special-config-storage.adoc:136]: s/the instructions above./the preceding instructions./ (IBMSG again)

File: modules/installation-special-config-storage.adoc [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+<7> Include this section if you want to mirror the boot disk. See "About disk mirroring" for more details.
+<8> List all disk devices that should be included in the boot disk mirror, including the disk that {op-system} will be installed onto.
+<9> Include this directive to enable FIPS mode on your cluster.
++
+[IMPORTANT]
+====
+If you are configuring nodes to use both disk encryption and mirroring, both features must be configured in the same Butane config.  In addition, if you are configuring disk encryption on a node with FIPS mode enabled, you must include the `fips` directive in the same Butane config, even if FIPS mode is also enabled in a separate manifest.
+====
+
+. In the `openshift` directory, use Butane to create a master or worker manifest from the Butane config. For example, to create a manifest for worker nodes, run the following command:
```
Comment (bobfuru) [inline modules/installation-special-config-storage.adoc:147]: s/master/control plane

**Review by bobfuru — COMMENTED** [review]
Excellent work, as always @bgilbert. A few small nits and suggestion for updating master -> control plane. Overall, this LGTM. LMK when you're ready for a merge. 🎉

File: modules/installation-special-config-storage.adoc [hunk 7]
```diff
@@ -0,0 +1,156 @@
...
+
+To enable boot disk encryption or mirroring during {product-title} deployment:
+
+. If you want to use TPM v2 to encrypt your cluster, check to see if TPM v2 encryption needs to be enabled in the BIOS on each node. This is required on most Dell systems. Check the manual for your computer.
+
+. If you want to use Tang to encrypt your cluster, follow these preparatory steps:
+
+.. Access a Red Hat Enterprise Linux server from which you can configure the encryption settings.
+
+.. Set up or access an existing Tang server. See link:https://access.redhat.com/documentation/en-us/red_hat_enterprise_linux/8/html/security_hardening/configuring-automated-unlocking-of-encrypted-volumes-using-policy-based-decryption_security-hardening#network-bound-disk-encryption_configuring-automated-unlocking-of-encrypted-volumes-using-policy-based-decryption[Network-bound disk encryption] for instructions. See link:https://youtu.be/2uLKvB8Z5D0[Securing Automated Decryption New Cryptography and Techniques] for a presentation on Tang.
```
[the remark this answers is not in the document]
Comment (bgilbert) [reply modules/installation-special-config-storage.adoc:57]: This was carried over verbatim from the [existing docs](https://docs.openshift.com/container-platform/4.7/installing/install_config/installing-customizing.html#installation-special-config-encrypt-disk-tang_installing-customizing), originally added in #19144.  I don't have an opinion on keeping the link; lmk if you'd like it removed.

**Comment by bgilbert** [comment]
Updated, and ready for merge unless https://github.com/openshift/openshift-docs/pull/33337#discussion_r653001524 should be addressed.

File: modules/installation-special-config-storage.adoc (hunk 7 above)
[the remark this answers is not in the document]
Comment (bobfuru) [reply modules/installation-special-config-storage.adoc:57]: In that case, let's keep it! 😆

**Pull request merged**